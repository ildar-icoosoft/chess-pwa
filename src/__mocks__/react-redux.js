const reactRedux = jest.genMockFromModule("react-redux");

reactRedux.useDispatch.mockReturnValue(jest.fn());

module.exports = reactRedux;
