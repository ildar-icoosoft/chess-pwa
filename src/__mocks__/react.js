import React from "react";

React.useEffect = jest.fn();

module.exports = React;
