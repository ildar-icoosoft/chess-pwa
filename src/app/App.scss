.container-fluid {
  width: 1140px;
}
