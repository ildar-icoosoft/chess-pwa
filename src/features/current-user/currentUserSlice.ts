/* eslint-disable no-param-reassign */
/* eslint-disable import/no-cycle */

import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { normalize } from "normalizr";
import { JWR } from "sails.io.js";
import { AppThunk } from "../../app/store";
import ioClient from "../../services/ioClient";
import User from "../../interfaces/User";
import LoginData from "../../interfaces/LoginData";
import SignUpData from "../../interfaces/SignUpData";
import userSchema from "../../normalizr/schemas/userSchema";
import NormalizedData from "../../normalizr/interfaces/NormalizedData";
import getErrorMessageFromJWR from "../../utils/getErrorMessageFromJWR";

interface CurrentUserState {
  userId: number | null;
  isLoading: boolean;
  error: string | null;
}

const initialState: CurrentUserState = {
  userId: null,
  isLoading: true,
  error: null,
};

const currentUserSlice = createSlice({
  name: "currentUser",
  initialState,
  reducers: {
    getCurrentUserRequest(state) {
      state.isLoading = true;
      state.error = null;
    },
    getCurrentUserSuccess(
      state,
      action: PayloadAction<NormalizedData<number> | null>
    ) {
      state.userId = action.payload ? action.payload.result : null;
      state.isLoading = false;
      state.error = null;
    },
    getCurrentUserError(state, action: PayloadAction<string>) {
      state.isLoading = false;
      state.error = action.payload;
    },
    loginRequest(_state) {},
    loginSuccess(state, action: PayloadAction<NormalizedData<number>>) {
      state.userId = action.payload.result;
    },
    loginError(_state, _action: PayloadAction<string>) {},
    registerRequest(_state) {},
    registerSuccess(state, action: PayloadAction<NormalizedData<number>>) {
      state.userId = action.payload.result;
    },
    registerError(_state, _action: PayloadAction<string>) {},
    logoutRequest(_state) {},
    logoutSuccess(state) {
      state.userId = null;
    },
    logoutError(_state, _action: PayloadAction<string>) {},
  },
});

export const {
  getCurrentUserRequest,
  getCurrentUserSuccess,
  getCurrentUserError,
  loginRequest,
  loginSuccess,
  loginError,
  registerRequest,
  registerSuccess,
  registerError,
  logoutRequest,
  logoutSuccess,
  logoutError,
} = currentUserSlice.actions;

export default currentUserSlice.reducer;

export const fetchCurrentUser = (): AppThunk<Promise<User | null>> => (
  dispatch
) => {
  dispatch(getCurrentUserRequest());

  return new Promise((resolve, reject) => {
    ioClient.socket.get("/api/v1/account/me", (body: unknown, jwr: JWR) => {
      if (jwr.statusCode === 200) {
        const normalizedUser = normalize(body as User, userSchema);

        dispatch(getCurrentUserSuccess(normalizedUser));
        resolve(body as User);
      } else if (jwr.statusCode === 401) {
        dispatch(getCurrentUserSuccess(null));
        resolve(null);
      } else {
        dispatch(getCurrentUserError(getErrorMessageFromJWR(jwr)));
        reject(jwr);
      }
    });
  });
};

export const login = (data: LoginData): AppThunk<Promise<User>> => (
  dispatch
) => {
  dispatch(loginRequest());

  return new Promise((resolve, reject) => {
    ioClient.socket.put(
      "/api/v1/entrance/login",
      {
        // rememberMe: true, // because of warning in backend @see https://tracker.yandex.ru/CHESSPWA-47
        emailAddress: data.email,
        password: data.password,
      },
      (body: unknown, jwr: JWR) => {
        if (jwr.statusCode === 200) {
          const normalizedUser = normalize(body as User, userSchema);

          dispatch(loginSuccess(normalizedUser));
          resolve(body as User);
        } else {
          dispatch(loginError(getErrorMessageFromJWR(jwr)));
          reject(jwr);
        }
      }
    );
  });
};

export const register = (data: SignUpData): AppThunk<Promise<User>> => (
  dispatch
) => {
  dispatch(registerRequest());

  return new Promise((resolve, reject) => {
    ioClient.socket.post(
      "/api/v1/entrance/signup",
      {
        fullName: data.fullName,
        emailAddress: data.email,
        password: data.password,
        confirmPassword: data.password,
        agreed: true,
      },
      (body: unknown, jwr: JWR) => {
        if (jwr.statusCode === 200) {
          const normalizedUser = normalize(body as User, userSchema);

          dispatch(registerSuccess(normalizedUser));
          resolve(body as User);
        } else {
          dispatch(registerError(getErrorMessageFromJWR(jwr)));
          reject(jwr);
        }
      }
    );
  });
};

export const logout = (): AppThunk<Promise<void>> => (dispatch) => {
  dispatch(logoutRequest());

  return new Promise((resolve, reject) => {
    ioClient.socket.post(
      "/api/v1/account/logout",
      {},
      (body: unknown, jwr: JWR) => {
        if (jwr.statusCode === 200) {
          dispatch(logoutSuccess());
          resolve();
        } else {
          dispatch(logoutError(getErrorMessageFromJWR(jwr)));
          reject(jwr);
        }
      }
    );
  });
};
