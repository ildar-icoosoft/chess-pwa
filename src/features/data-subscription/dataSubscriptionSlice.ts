/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-empty-interface */
/* eslint-disable import/no-cycle */

import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { normalize } from "normalizr";
import NormalizedData from "../../normalizr/interfaces/NormalizedData";
import { AppThunk } from "../../app/store";
import ioClient from "../../services/ioClient";
import { SubscriptionData } from "../../interfaces/SubscriptionData";
import gameSchema from "../../normalizr/schemas/gameSchema";
import seekSchema from "../../normalizr/schemas/seekSchema";
import userSchema from "../../normalizr/schemas/userSchema";
import chatMessageSchema from "../../normalizr/schemas/chatMessageSchema";

interface DataSubscriptionState {}

const initialState: DataSubscriptionState = {};

const dataSubscriptionSlice = createSlice({
  name: "dataSubscription",
  initialState,
  reducers: {
    updateGameBySubscription(
      _state,
      _action: PayloadAction<NormalizedData<number>>
    ) {},
    createGameBySubscription(
      _state,
      _action: PayloadAction<NormalizedData<number>>
    ) {},
    updateSeekBySubscription(
      _state,
      _action: PayloadAction<NormalizedData<number>>
    ) {},
    createSeekBySubscription(
      _state,
      _action: PayloadAction<NormalizedData<number>>
    ) {},
    removeSeekBySubscription(_state, _action: PayloadAction<number>) {},
    createUserBySubscription(
      _state,
      _action: PayloadAction<NormalizedData<number>>
    ) {},
    updateUserBySubscription(
      _state,
      _action: PayloadAction<NormalizedData<number>>
    ) {},
    createChatMessageBySubscription(
      _state,
      _action: PayloadAction<NormalizedData<number>>
    ) {},
    disconnectSocket(_state) {},
    reconnectSocket(_state) {},
  },
  extraReducers: {},
});

export const {
  updateGameBySubscription,
  createGameBySubscription,
  updateSeekBySubscription,
  createSeekBySubscription,
  removeSeekBySubscription,
  createUserBySubscription,
  updateUserBySubscription,
  createChatMessageBySubscription,
  reconnectSocket,
  disconnectSocket,
} = dataSubscriptionSlice.actions;

export default dataSubscriptionSlice.reducer;

export const watchGames = (): AppThunk<void> => (dispatch) => {
  ioClient.socket.on("game", (subscriptionData: SubscriptionData) => {
    if (subscriptionData.verb === "updated") {
      const game = {
        ...subscriptionData.previous,
        ...subscriptionData.data,
      };

      const normalizedGame = normalize(game, gameSchema);

      dispatch(updateGameBySubscription(normalizedGame));
    } else if (subscriptionData.verb === "created") {
      const normalizedGame = normalize(subscriptionData.data, gameSchema);

      dispatch(createGameBySubscription(normalizedGame));
    }
  });
};

export const watchSeeks = (): AppThunk<void> => (dispatch) => {
  ioClient.socket.on("seek", (subscriptionData: SubscriptionData) => {
    if (subscriptionData.verb === "updated") {
      const seek = {
        ...subscriptionData.previous,
        ...subscriptionData.data,
      };

      const normalizedSeek = normalize(seek, seekSchema);

      dispatch(updateSeekBySubscription(normalizedSeek));
    } else if (subscriptionData.verb === "created") {
      const normalizedSeek = normalize(subscriptionData.data, seekSchema);

      dispatch(createSeekBySubscription(normalizedSeek));
    } else if (subscriptionData.verb === "destroyed") {
      dispatch(removeSeekBySubscription(subscriptionData.id));
    }
  });
};

export const watchUsers = (): AppThunk<void> => (dispatch) => {
  ioClient.socket.on("user", (subscriptionData: SubscriptionData) => {
    if (subscriptionData.verb === "updated") {
      const seek = {
        ...subscriptionData.previous,
        ...subscriptionData.data,
      };

      const normalizedSeek = normalize(seek, userSchema);

      dispatch(updateUserBySubscription(normalizedSeek));
    } else if (subscriptionData.verb === "created") {
      const normalizedSeek = normalize(subscriptionData.data, userSchema);

      dispatch(createUserBySubscription(normalizedSeek));
    }
  });
};

export const watchChatMessages = (): AppThunk<void> => (dispatch) => {
  ioClient.socket.on("chat-message", (subscriptionData: SubscriptionData) => {
    if (subscriptionData.verb === "created") {
      const normalizedChatMessage = normalize(
        subscriptionData.data,
        chatMessageSchema
      );

      dispatch(createChatMessageBySubscription(normalizedChatMessage));
    }
  });
};

export const watchConnection = (): AppThunk<void> => (dispatch) => {
  ioClient.socket.on("disconnect", () => {
    dispatch(disconnectSocket());
  });

  ioClient.socket.on("reconnect", () => {
    dispatch(reconnectSocket());

    setTimeout(() => {
      document.location.reload();
    }, 3000);
  });
};
