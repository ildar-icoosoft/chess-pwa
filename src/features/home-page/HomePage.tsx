import React, { FC } from "react";
import { Button, Modal } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import OngoingGamesContainer from "../ongoing-games/OngoingGamesContainer";
import ChallengeAiFormContainer from "../challenge-ai-modal/ChallengeAiFormContainer";

import { RootState } from "../../app/rootReducer";
import {
  hideChallengeAiModal,
  showChallengeAiModal,
} from "../challenge-ai-modal/challengeAiModalSlice";
import { AppDispatch } from "../../app/store";

const HomePage: FC<unknown> = () => {
  const dispatch = useDispatch<AppDispatch>();

  const { isChallengeAiModalVisible } = useSelector(
    (state: RootState) => state.challengeAiModal
  );

  return (
    <>
      <Modal
        show={isChallengeAiModalVisible}
        onHide={() => dispatch(hideChallengeAiModal())}
        animation={false}
      >
        <Modal.Header closeButton>
          <Modal.Title>Play with the computer</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <ChallengeAiFormContainer />
        </Modal.Body>
      </Modal>
      <Button
        variant="primary"
        onClick={() => dispatch(showChallengeAiModal())}
      >
        Play with the computer
      </Button>
      <OngoingGamesContainer />
    </>
  );
};

export default HomePage;
