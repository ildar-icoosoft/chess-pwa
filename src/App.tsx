import React, { FC, useCallback, useState } from "react";
import { BrowserRouter as Router, Switch, Route, Link } from "react-router-dom";
import "bootstrap/dist/css/bootstrap.min.css";
import HomePage from "./pages/HomePage";
import GamePage from "./pages/GamePage";
import { LoginForm } from "./components/LoginForm";
import { Button, Modal } from "react-bootstrap";

const App: FC = () => {
  const [modalIsVisible, setModalVisibility] = useState(false);

  const showModal = useCallback(() => {
    setModalVisibility(true);
  }, []);
  const hideModal = useCallback(() => {
    setModalVisibility(false);
  }, []);

  return (
    <Router>
      <Button variant="primary" onClick={showModal}>
        Login / Register
      </Button>
      <Modal show={modalIsVisible} onHide={hideModal} animation={false}>
        <Modal.Header closeButton>
          <Modal.Title>Login</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <LoginForm />
        </Modal.Body>
      </Modal>
      <div>
        <nav>
          <ul>
            <li>
              <Link to="/">Home</Link>
            </li>
          </ul>
        </nav>

        <Switch>
          <Route path="/game/:id">
            <GamePage />
          </Route>
          <Route path="/">
            <HomePage />
          </Route>
        </Switch>
      </div>
    </Router>
  );
};

export default App;
