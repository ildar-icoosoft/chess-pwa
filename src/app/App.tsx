import React, { FC, useEffect } from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { useDispatch } from "react-redux";
import "bootstrap/dist/css/bootstrap.min.css";
import HomePage from "../features/home-page/HomePage";
import GamePage from "../pages/GamePage";
import { fetchCurrentUser } from "../features/current-user/currentUserSlice";
import { watchGames } from "../features/data-subscription/dataSubscriptionSlice";
import { startGameClock } from "../features/game-clock/gameClockSlice";
import HeaderContainer from "../features/header/HeaderContainer";
import AuthModalContainer from "../features/auth-modal/AuthModalContainer";

const App: FC = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCurrentUser());
    dispatch(watchGames());
    dispatch(startGameClock());
  }, [dispatch]);

  return (
    <div className="container">
      <Router>
        <HeaderContainer />
        <AuthModalContainer />

        <Switch>
          <Route path="/game/:id">
            <GamePage />
          </Route>
          <Route path="/">
            <HomePage />
          </Route>
        </Switch>
      </Router>
    </div>
  );
};

export default App;
