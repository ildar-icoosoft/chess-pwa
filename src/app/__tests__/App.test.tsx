import React from "react";
import TestRenderer from "react-test-renderer";
import { useDispatch, useSelector } from "react-redux";
import { render } from "@testing-library/react";
import App from "../App";
import mountTest from "../../tests/mountTest";
import HomePage from "../../pages/HomePage";
import { fetchCurrentUser } from "../../redux/slices/currentUserSlice";
import { watchGames } from "../../redux/slices/dataSubscriptionSlice";
import { defaultState } from "../../tests/data-sample/state";

jest.useFakeTimers();

jest.mock("../../redux/slices/currentUserSlice");
jest.mock("../../redux/slices/dataSubscriptionSlice");

describe("App", () => {
  beforeAll(() => {
    (useSelector as jest.Mock).mockImplementation((cb) => cb(defaultState));
  });
  beforeEach(() => {
    useDispatch<jest.Mock>().mockClear();
  });

  mountTest(App);

  it("Snapshot", () => {
    const tree = TestRenderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("children components", () => {
    it("contains HomePage", () => {
      const testRenderer = TestRenderer.create(<App />);
      const testInstance = testRenderer.root;

      expect(testInstance.findAllByType(HomePage).length).toBe(1);
    });
  });

  // @todo. add tests for App contains HomePage and GamePage
  // @todo. add tests for other content
  // @todo. add tests for auth modal
  // @todo. add tests for AppContext

  describe("DOM structure", () => {
    it("renders learn react link", () => {
      const { getByText } = render(<App />);
      const linkElement = getByText(/Home/i);
      expect(linkElement).toBeInTheDocument();
    });
  });

  describe("dispatch() calls", () => {
    it("should call dispatch(fetchCurrentUser())", () => {
      const dispatch = useDispatch<jest.Mock>();

      const fetchCurrentUserReturnedValue = Symbol();

      const fetchCurrentUserFn = fetchCurrentUser as jest.Mock;
      fetchCurrentUserFn.mockClear();
      fetchCurrentUserFn.mockReturnValue(fetchCurrentUserReturnedValue);

      TestRenderer.act(() => {
        TestRenderer.create(<App />);
      });

      expect(fetchCurrentUserFn).toBeCalledTimes(1);
      expect(fetchCurrentUserFn).toBeCalledWith();

      expect(dispatch).toBeCalledWith(fetchCurrentUserReturnedValue);
    });

    it("watchGames()", () => {
      const dispatch = useDispatch<jest.Mock>();

      const watchGamesReturnedValue = Symbol();

      const watchGamesFn = watchGames as jest.Mock;
      watchGamesFn.mockClear();
      watchGamesFn.mockReturnValue(watchGamesReturnedValue);

      TestRenderer.act(() => {
        TestRenderer.create(<App />);
      });

      expect(watchGamesFn).toBeCalledTimes(1);
      expect(watchGamesFn).toBeCalledWith();

      expect(dispatch).toBeCalledWith(watchGamesReturnedValue);
    });
  });
});
