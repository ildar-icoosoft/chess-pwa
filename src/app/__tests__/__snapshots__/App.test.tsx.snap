// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Snapshot 1`] = `
<div
  className="container-fluid"
>
  <header
    className="d-flex flex-column flex-md-row align-items-center p-3 px-md-4 mb-3 bg-white border-bottom box-shadow"
  >
    <div
      className="offset-4 col-4 text-center"
    >
      Chess PWA
    </div>
    <div
      className="col-4 d-flex justify-content-end align-items-center"
    >
      <button
        className="btn btn-sm btn-success mr-3 btn btn-primary"
        data-testid="login-btn"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Login / Register
      </button>
      <a
        href="https://github.com/ildar-icoosoft/chess-pwa"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt=""
          src="/imgs/GitHub-Mark-32px.png"
        />
      </a>
    </div>
  </header>
  <div />
  <div
    className="messages"
  />
  <div
    className="d-flex justify-content-center"
  >
    <div
      className="mb-3"
    >
      <button
        className="mr-3 btn btn-primary"
        data-testid="create-game-btn"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Create a game
      </button>
      <button
        className="btn btn-primary"
        data-testid="challenge-ai-btn"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Play with the computer
      </button>
    </div>
  </div>
  <div
    className="mb-3 card"
  >
    <div
      className="card-body"
    >
      <div
        className="card-title h5"
      >
        Waiting for opponent
      </div>
      <div>
        <div
          className="text-center"
        >
          <div
            className="spinner-border"
            data-testid="loading-spinner"
            role="status"
          >
            <span
              className="sr-only"
            >
              Loading...
            </span>
          </div>
        </div>
        <div />
      </div>
    </div>
  </div>
  <div
    className="mb-3 card"
  >
    <div
      className="card-body"
    >
      <div
        className="card-title h5"
      >
        Playing right now
      </div>
      <div>
        <div
          className="text-center"
        >
          <div
            className="spinner-border"
            data-testid="loading-spinner"
            role="status"
          >
            <span
              className="sr-only"
            >
              Loading...
            </span>
          </div>
        </div>
        <div
          className="grid"
        />
      </div>
    </div>
  </div>
  <div
    className="card"
  >
    <div
      className="card-body"
    >
      <div
        className="card-title h5"
      >
        Finished
      </div>
      <div>
        <div
          className="text-center"
        >
          <div
            className="spinner-border"
            data-testid="loading-spinner"
            role="status"
          >
            <span
              className="sr-only"
            >
              Loading...
            </span>
          </div>
        </div>
        <div
          className="grid"
        />
      </div>
    </div>
  </div>
</div>
`;
